pkg/controllers/migration/controller.go

```diff
@@ -0,0 +1,117 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package migration
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"strings"
+
+	"github.com/awslabs/operatorpkg/object"
+	"github.com/awslabs/operatorpkg/singleton"
+	"github.com/samber/lo"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	v1 "sigs.k8s.io/karpenter/pkg/apis/v1"
+	"sigs.k8s.io/karpenter/pkg/operator/injection"
+)
+
+type Controller[T client.Object] struct {
+	kubeClient client.Client
+}
+
+func NewController[T client.Object](client client.Client) *Controller[T] {
+	return &Controller[T]{
+		kubeClient: client,
+	}
+}
+
+func (c *Controller[T]) Reconcile(ctx context.Context) (reconcile.Result, error) {
+	ctx = injection.WithControllerName(ctx, "migration")
+
+	// create object list
+	o := object.New[T]()
+	list := &unstructured.UnstructuredList{}
+	list.SetGroupVersionKind(object.GVK(o))
+	if err := c.kubeClient.List(ctx, list); err != nil {
+		return reconcile.Result{}, fmt.Errorf("listing objects, %w", err)
+	}
+
+	// update annotations on all CRs
+	for _, item := range list.Items {
+		if lo.HasKey(item.GetAnnotations(), v1.StoredVersionMigrated) {
+			continue
+		}
+		item.SetAnnotations(lo.Assign(item.GetAnnotations(), map[string]string{
+			v1.StoredVersionMigrated: "true",
+		}))
+		if err := c.kubeClient.Update(ctx, &item); err != nil {
+			return reconcile.Result{}, fmt.Errorf("annotating: %w", err)
+		}
+	}
+
+	// To get here, there shouldn't be any resources missing the annotation so filter for and update CRD
+	crdList := &apiextensionsv1.CustomResourceDefinitionList{}
+	if err := c.kubeClient.List(ctx, crdList); err != nil {
+		return reconcile.Result{}, fmt.Errorf("getting crds, %w", err)
+	}
+	// should filter tighter than this
+	crds := lo.Filter(crdList.Items, func(crd apiextensionsv1.CustomResourceDefinition, _ int) bool {
+		return strings.Contains(crd.ObjectMeta.Name, strings.ToLower(object.GVK(o).Kind))
+	})
+	for _, crd := range crds {
+		stored := crd.DeepCopy()
+		crd.Status.StoredVersions = []string{"v1"}
+		if err := c.kubeClient.Status().Patch(ctx, &crd, client.MergeFromWithOptions(stored, client.MergeFromWithOptimisticLock{})); err != nil {
+			return reconcile.Result{}, fmt.Errorf("updating crd: %w", err)
+		}
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (c *Controller[T]) Register(_ context.Context, m manager.Manager) error {
+	o := object.New[T]()
+	return controllerruntime.NewControllerManagedBy(m).
+		Named(fmt.Sprintf("migration.%s", strings.ToLower(reflect.TypeOf(o).Elem().Name()))).
+		For(o).
```

Comment:
Are we reconciling against the CRs or the CRDs? This `For` sets us to reconcile against the CRs, but the following `Watches` and the structure of the reconcile loop itself suggests we're reconciling against CRDs. The latter seems correct to me.

Reply:
Seems like we could implement this in a single controller rather than 3 typed controllers. We could use an EventFilter to only look at CRDs which are part of one of our groups, and have a v1 storage version.

Reply:
Only issue is the NodeClass, right? Also, that introduces the need for a bunch of filtering for multiple CRs and then updating the CRD for each based on the state of CRs.

Reply:
IIUC this watches the CRDs so that it can list it from the shared informer and not as a hard list.